### DISCUSSION TITLE
How to reduce the size of an image

### DISCUSSION THREAD

**Discussion by alSeveR** [pymupdf/PyMuPDF#1335] [Q&A] [answered]
I'm testing this code, but it doesn't matter what values you put in set_dpi. As a test image I am using this one: https://www.pngarts.com/files/2/Orange-Abstract-Lines-Transparent-Background-PNG.png
```
doc = fitz.open("test.pdf")
stream = open("test.png", "rb").read()
pix = fitz.Pixmap(stream)
pix.set_dpi(150, 150)
rect = fitz.Rect(15, 100, 65, 165)
for page in doc:
    page.insert_image(rect, pixmap=pix, overlay=True)
doc.save("out.pdf", incremental=False, encryption=0)
doc.close()
```
Thank you in advance.

**Comment by JorjMcKie** [comment]
What do you want to achieve: 

- [ ] a small file size of the image embedded via ``insert_image``?
- [ ] Or a small display on the page?
- [ ] Or the image physically inserted only once, but shown on every page?

With all of the above, the image resolution has nothing to do.

**Comment by JorjMcKie** [comment]
It seems you want to give up image quality for the sake of inserted image size.
You must "shrink" the pixmap for this - there exists a separate pixmap method.
Or make a pixmap copy, setting new width and height values - see documentation about possible pixmap constructors.

**Comment by alSeveR** [comment]
>  a small file size of the image embedded via insert_im age?

Yes.
> Or a small display on the page?

No. I think that I can control it with rect.
> Or the image physically inserted only once, but shown on every page?

It is pissible and the result pdf is more lightweight?

I have tried using shrink, but the minimum reduction is half and, in some cases, it is too much.

I will have a look at the documentation about pixmap constructors.

Thanks.

**Comment by JorjMcKie** [comment, marked as answer]
If you insert the same image on multiple pages the logic automatically determines this and reuses the already embedded image.
It is however much faster to avoid the effort detecting this and use this kind of loop:
```python
xref = 0  # xref start value
for page in doc:
    xref = page.insert_image(rect, ..., xref=xref)  # will pick up any previously returned xref > 0
```
This is how to scale down continuously:
![grafik](https://user-images.githubusercontent.com/8290722/138499463-48fe170c-abfa-4f9c-a24c-1c5d3adf5b65.png)

At the end, do not forget to save the document with garbage>=3, deflate=True

**Reply by JorjMcKie** [reply]
of course you can also use Pillow method `.thumbnail()` and insert the image outcome of this ...

**Comment by alSeveR** [comment]
Thank you very much. It works like a charm.

**Reply by JorjMcKie** [reply]
you are welcome
